<a href="https://colab.research.google.com/github/pranay8297/deep-learning-projects/blob/master/week_6_assignment_complete.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Notebook Walkthrough

In [1]:
import nltk
nltk.download('brown')  

[nltk_data] Downloading package brown to /root/nltk_data...
[nltk_data]   Unzipping corpora/brown.zip.


True

In [3]:
from nltk.corpus import brown
brown.words() # Returns a list of strings

['The', 'Fulton', 'County', 'Grand', 'Jury', 'said', ...]

In [117]:
from nltk.corpus import gutenberg
nltk.download('gutenberg')

[nltk_data] Downloading package gutenberg to /root/nltk_data...
[nltk_data]   Unzipping corpora/gutenberg.zip.


True

In [118]:
len(gutenberg.sents()), gutenberg.sents()[0]

(98552, ['[', 'Emma', 'by', 'Jane', 'Austen', '1816', ']'])

In [121]:
gutenberg.fileids()[:4], len(gutenberg.fileids())

(['austen-emma.txt',
  'austen-persuasion.txt',
  'austen-sense.txt',
  'bible-kjv.txt'],
 18)

In [122]:
gutenberg.sents(fileids='austen-emma.txt') , len(gutenberg.sents(fileids='austen-emma.txt'))

([['[', 'Emma', 'by', 'Jane', 'Austen', '1816', ']'], ['VOLUME', 'I'], ...],
 7752)

In [124]:
gutenberg.raw('bible-kjv.txt')[:400]

'[The King James Bible]\n\nThe Old Testament of the King James Bible\n\nThe First Book of Moses:  Called Genesis\n\n\n1:1 In the beginning God created the heaven and the earth.\n\n1:2 And the earth was without form, and void; and darkness was upon\nthe face of the deep. And the Spirit of God moved upon the face of the\nwaters.\n\n1:3 And God said, Let there be light: and there was light.\n\n1:4 And God saw the li'

In [133]:
x = gutenberg.raw('bible-kjv.txt')
sentences = x.split('\n\n')
sentences[:20]

['[The King James Bible]',
 'The Old Testament of the King James Bible',
 'The First Book of Moses:  Called Genesis',
 '\n1:1 In the beginning God created the heaven and the earth.',
 '1:2 And the earth was without form, and void; and darkness was upon\nthe face of the deep. And the Spirit of God moved upon the face of the\nwaters.',
 '1:3 And God said, Let there be light: and there was light.',
 '1:4 And God saw the light, that it was good: and God divided the light\nfrom the darkness.',
 '1:5 And God called the light Day, and the darkness he called Night.\nAnd the evening and the morning were the first day.',
 '1:6 And God said, Let there be a firmament in the midst of the waters,\nand let it divide the waters from the waters.',
 '1:7 And God made the firmament, and divided the waters which were\nunder the firmament from the waters which were above the firmament:\nand it was so.',
 '1:8 And God called the firmament Heaven. And the evening and the\nmorning were the second day.',
 '1:9

In [136]:
s8 = sentences[16]
s8

'1:14 And God said, Let there be lights in the firmament of the heaven\nto divide the day from the night; and let them be for signs, and for\nseasons, and for days, and years: 1:15 And let them be for lights in\nthe firmament of the heaven to give light upon the earth: and it was\nso.'

In [137]:
from nltk import sent_tokenize, word_tokenize
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [138]:
sent_tokenize(s8)

['1:14 And God said, Let there be lights in the firmament of the heaven\nto divide the day from the night; and let them be for signs, and for\nseasons, and for days, and years: 1:15 And let them be for lights in\nthe firmament of the heaven to give light upon the earth: and it was\nso.']

In [139]:
for sent in sent_tokenize(s8):
    print(word_tokenize(sent))

['1:14', 'And', 'God', 'said', ',', 'Let', 'there', 'be', 'lights', 'in', 'the', 'firmament', 'of', 'the', 'heaven', 'to', 'divide', 'the', 'day', 'from', 'the', 'night', ';', 'and', 'let', 'them', 'be', 'for', 'signs', ',', 'and', 'for', 'seasons', ',', 'and', 'for', 'days', ',', 'and', 'years', ':', '1:15', 'And', 'let', 'them', 'be', 'for', 'lights', 'in', 'the', 'firmament', 'of', 'the', 'heaven', 'to', 'give', 'light', 'upon', 'the', 'earth', ':', 'and', 'it', 'was', 'so', '.']


In [140]:
for sent in sent_tokenize(s8):
    # It's a little in efficient to loop through each word,
    # after but sometimes it helps to get better tokens.
    print([word.lower() for word in word_tokenize(sent)])
    # Alternatively:
    #print(list(map(str.lower, word_tokenize(sent))))

['1:14', 'and', 'god', 'said', ',', 'let', 'there', 'be', 'lights', 'in', 'the', 'firmament', 'of', 'the', 'heaven', 'to', 'divide', 'the', 'day', 'from', 'the', 'night', ';', 'and', 'let', 'them', 'be', 'for', 'signs', ',', 'and', 'for', 'seasons', ',', 'and', 'for', 'days', ',', 'and', 'years', ':', '1:15', 'and', 'let', 'them', 'be', 'for', 'lights', 'in', 'the', 'firmament', 'of', 'the', 'heaven', 'to', 'give', 'light', 'upon', 'the', 'earth', ':', 'and', 'it', 'was', 'so', '.']


In [141]:
print(word_tokenize(s8))

['1:14', 'And', 'God', 'said', ',', 'Let', 'there', 'be', 'lights', 'in', 'the', 'firmament', 'of', 'the', 'heaven', 'to', 'divide', 'the', 'day', 'from', 'the', 'night', ';', 'and', 'let', 'them', 'be', 'for', 'signs', ',', 'and', 'for', 'seasons', ',', 'and', 'for', 'days', ',', 'and', 'years', ':', '1:15', 'And', 'let', 'them', 'be', 'for', 'lights', 'in', 'the', 'firmament', 'of', 'the', 'heaven', 'to', 'give', 'light', 'upon', 'the', 'earth', ':', 'and', 'it', 'was', 'so', '.']


In [143]:
single_no9 = webtext.raw('singles.txt').split('\n')[17]
sent_tokenize(single_no9)

['BLONDE HAIR, BLUE EYES Medium build, Im in my early 30s.',
 'Am honest, caring, likes fishing, animals, golf, bike riding, TV and DVDs, quiet nights at home, kissing and cuddling up to a special person in my life.',
 'Looking for a caring, honest lady for friendship to relationship.']

In [144]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [145]:
from nltk.corpus import stopwords

stopwords_en = stopwords.words('english')
print(stopwords_en)

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '

In [146]:
single_no8_tokenized_lowered = list(map(str.lower, word_tokenize(s8)))
print(single_no8_tokenized_lowered)

['1:14', 'and', 'god', 'said', ',', 'let', 'there', 'be', 'lights', 'in', 'the', 'firmament', 'of', 'the', 'heaven', 'to', 'divide', 'the', 'day', 'from', 'the', 'night', ';', 'and', 'let', 'them', 'be', 'for', 'signs', ',', 'and', 'for', 'seasons', ',', 'and', 'for', 'days', ',', 'and', 'years', ':', '1:15', 'and', 'let', 'them', 'be', 'for', 'lights', 'in', 'the', 'firmament', 'of', 'the', 'heaven', 'to', 'give', 'light', 'upon', 'the', 'earth', ':', 'and', 'it', 'was', 'so', '.']


In [147]:
stopwords_en = set(stopwords.words('english')) # Set checking is faster in Python than list.

# List comprehension.
print([word for word in single_no8_tokenized_lowered if word not in stopwords_en])

['1:14', 'god', 'said', ',', 'let', 'lights', 'firmament', 'heaven', 'divide', 'day', 'night', ';', 'let', 'signs', ',', 'seasons', ',', 'days', ',', 'years', ':', '1:15', 'let', 'lights', 'firmament', 'heaven', 'give', 'light', 'upon', 'earth', ':', '.']


In [148]:
from string import punctuation
# It's a string so we have to them into a set type
print('From string.punctuation:', type(punctuation), punctuation)

From string.punctuation: <class 'str'> !"#$%&'()*+,-./:;<=>?@[\]^_`{|}~


In [149]:
stopwords_en_withpunct = stopwords_en.union(set(punctuation))
print(stopwords_en_withpunct)

{'few', "mightn't", "'", 'those', 'or', 'there', 'her', 'too', 'very', "shouldn't", '[', 'against', "aren't", 'haven', "wouldn't", 'to', 'myself', 'he', 'doing', '@', 'some', 'here', 'wasn', 'other', 'out', '.', 'which', 'll', "she's", 'itself', "doesn't", 'most', 'than', 'don', 'are', '+', 'and', 'down', 'mightn', ':', 'who', "isn't", 'after', "hadn't", 'won', 'did', 'am', 'how', '$', 'ours', 'ourselves', 'was', 'they', 'can', 'these', "that'll", 'so', '%', 'aren', 'has', "haven't", "it's", 'should', 'you', 'as', '/', 'their', 'just', 'now', 'does', 've', '&', 'same', 'the', 'with', 's', 'into', 'had', 'in', "you're", 'for', "mustn't", ';', 'couldn', 'didn', '|', '#', '(', 'o', '}', 'more', "shan't", 'yourselves', 'during', 'all', ')', 'whom', 'yours', 'this', ',', 'of', 'that', 'hasn', 'himself', 'needn', 'up', 'once', 'wouldn', 'will', 'hers', 'each', "you'd", 'shan', 'off', '?', '\\', 'but', 'themselves', "didn't", 'have', 'shouldn', 'where', 'our', 'were', 'again', 'is', '`', "was

In [153]:
s8_clean_initail = print([word for word in single_no8_tokenized_lowered if word not in stopwords_en_withpunct])
s8_clean_initail

['1:14', 'god', 'said', 'let', 'lights', 'firmament', 'heaven', 'divide', 'day', 'night', 'let', 'signs', 'seasons', 'days', 'years', '1:15', 'let', 'lights', 'firmament', 'heaven', 'give', 'light', 'upon', 'earth']


In [154]:
stopwords_json = {"en":["a","a's","able","about","above","according","accordingly","across","actually","after","afterwards","again","against","ain't","all","allow","allows","almost","alone","along","already","also","although","always","am","among","amongst","an","and","another","any","anybody","anyhow","anyone","anything","anyway","anyways","anywhere","apart","appear","appreciate","appropriate","are","aren't","around","as","aside","ask","asking","associated","at","available","away","awfully","b","be","became","because","become","becomes","becoming","been","before","beforehand","behind","being","believe","below","beside","besides","best","better","between","beyond","both","brief","but","by","c","c'mon","c's","came","can","can't","cannot","cant","cause","causes","certain","certainly","changes","clearly","co","com","come","comes","concerning","consequently","consider","considering","contain","containing","contains","corresponding","could","couldn't","course","currently","d","definitely","described","despite","did","didn't","different","do","does","doesn't","doing","don't","done","down","downwards","during","e","each","edu","eg","eight","either","else","elsewhere","enough","entirely","especially","et","etc","even","ever","every","everybody","everyone","everything","everywhere","ex","exactly","example","except","f","far","few","fifth","first","five","followed","following","follows","for","former","formerly","forth","four","from","further","furthermore","g","get","gets","getting","given","gives","go","goes","going","gone","got","gotten","greetings","h","had","hadn't","happens","hardly","has","hasn't","have","haven't","having","he","he's","hello","help","hence","her","here","here's","hereafter","hereby","herein","hereupon","hers","herself","hi","him","himself","his","hither","hopefully","how","howbeit","however","i","i'd","i'll","i'm","i've","ie","if","ignored","immediate","in","inasmuch","inc","indeed","indicate","indicated","indicates","inner","insofar","instead","into","inward","is","isn't","it","it'd","it'll","it's","its","itself","j","just","k","keep","keeps","kept","know","known","knows","l","last","lately","later","latter","latterly","least","less","lest","let","let's","like","liked","likely","little","look","looking","looks","ltd","m","mainly","many","may","maybe","me","mean","meanwhile","merely","might","more","moreover","most","mostly","much","must","my","myself","n","name","namely","nd","near","nearly","necessary","need","needs","neither","never","nevertheless","new","next","nine","no","nobody","non","none","noone","nor","normally","not","nothing","novel","now","nowhere","o","obviously","of","off","often","oh","ok","okay","old","on","once","one","ones","only","onto","or","other","others","otherwise","ought","our","ours","ourselves","out","outside","over","overall","own","p","particular","particularly","per","perhaps","placed","please","plus","possible","presumably","probably","provides","q","que","quite","qv","r","rather","rd","re","really","reasonably","regarding","regardless","regards","relatively","respectively","right","s","said","same","saw","say","saying","says","second","secondly","see","seeing","seem","seemed","seeming","seems","seen","self","selves","sensible","sent","serious","seriously","seven","several","shall","she","should","shouldn't","since","six","so","some","somebody","somehow","someone","something","sometime","sometimes","somewhat","somewhere","soon","sorry","specified","specify","specifying","still","sub","such","sup","sure","t","t's","take","taken","tell","tends","th","than","thank","thanks","thanx","that","that's","thats","the","their","theirs","them","themselves","then","thence","there","there's","thereafter","thereby","therefore","therein","theres","thereupon","these","they","they'd","they'll","they're","they've","think","third","this","thorough","thoroughly","those","though","three","through","throughout","thru","thus","to","together","too","took","toward","towards","tried","tries","truly","try","trying","twice","two","u","un","under","unfortunately","unless","unlikely","until","unto","up","upon","us","use","used","useful","uses","using","usually","uucp","v","value","various","very","via","viz","vs","w","want","wants","was","wasn't","way","we","we'd","we'll","we're","we've","welcome","well","went","were","weren't","what","what's","whatever","when","whence","whenever","where","where's","whereafter","whereas","whereby","wherein","whereupon","wherever","whether","which","while","whither","who","who's","whoever","whole","whom","whose","why","will","willing","wish","with","within","without","won't","wonder","would","wouldn't","x","y","yes","yet","you","you'd","you'll","you're","you've","your","yours","yourself","yourselves","z","zero"]}
stopwords_json_en = set(stopwords_json['en'])
stopwords_nltk_en = set(stopwords.words('english'))
stopwords_punct = set(punctuation)
# Combine the stopwords. Its a lot longer so I'm not printing it out...
stoplist_combined = set.union(stopwords_json_en, stopwords_nltk_en, stopwords_punct)

# Remove the stopwords from `single_no8`.
print('With combined stopwords:')
print([word for word in single_no8_tokenized_lowered if word not in stoplist_combined])

With combined stopwords:
['1:14', 'god', 'lights', 'firmament', 'heaven', 'divide', 'day', 'night', 'signs', 'seasons', 'days', 'years', '1:15', 'lights', 'firmament', 'heaven', 'give', 'light', 'earth']


In [155]:
import nltk
nltk.download('wordnet')
import nltk
nltk.download('averaged_perceptron_tagger')
  

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


True

In [156]:
from nltk.stem import PorterStemmer
porter = PorterStemmer()

for word in ['walking', 'walks', 'walked']:
    print(porter.stem(word))

walk
walk
walk


In [158]:
from nltk.stem import WordNetLemmatizer
wnl = WordNetLemmatizer()

for word in ['walking', 'walks', 'walked']:
    print(wnl.lemmatize(word))

walking
walk
walked


In [160]:
from nltk import pos_tag
from nltk.stem import WordNetLemmatizer

wnl = WordNetLemmatizer()

def penn2morphy(penntag):
    """ Converts Penn Treebank tags to WordNet. """
    morphy_tag = {'NN':'n', 'JJ':'a',
                  'VB':'v', 'RB':'r'}
    try:
        return morphy_tag[penntag[:2]]
    except:
        return 'n' # if mapping isn't found, fall back to Noun.
    
# `pos_tag` takes the tokenized sentence as input, i.e. list of string,
# and returns a tuple of (word, tg), i.e. list of tuples of strings
# so we need to get the tag from the 2nd element.

walking_tagged = pos_tag(word_tokenize(s8))
print(walking_tagged)

[('1:14', 'CD'), ('And', 'CC'), ('God', 'NNP'), ('said', 'VBD'), (',', ','), ('Let', 'VB'), ('there', 'EX'), ('be', 'VB'), ('lights', 'NNS'), ('in', 'IN'), ('the', 'DT'), ('firmament', 'NN'), ('of', 'IN'), ('the', 'DT'), ('heaven', 'NN'), ('to', 'TO'), ('divide', 'VB'), ('the', 'DT'), ('day', 'NN'), ('from', 'IN'), ('the', 'DT'), ('night', 'NN'), (';', ':'), ('and', 'CC'), ('let', 'VB'), ('them', 'PRP'), ('be', 'VB'), ('for', 'IN'), ('signs', 'NNS'), (',', ','), ('and', 'CC'), ('for', 'IN'), ('seasons', 'NNS'), (',', ','), ('and', 'CC'), ('for', 'IN'), ('days', 'NNS'), (',', ','), ('and', 'CC'), ('years', 'NNS'), (':', ':'), ('1:15', 'CD'), ('And', 'CC'), ('let', 'VB'), ('them', 'PRP'), ('be', 'VB'), ('for', 'IN'), ('lights', 'NNS'), ('in', 'IN'), ('the', 'DT'), ('firmament', 'NN'), ('of', 'IN'), ('the', 'DT'), ('heaven', 'NN'), ('to', 'TO'), ('give', 'VB'), ('light', 'JJ'), ('upon', 'IN'), ('the', 'DT'), ('earth', 'NN'), (':', ':'), ('and', 'CC'), ('it', 'PRP'), ('was', 'VBD'), ('so',

In [162]:
print([wnl.lemmatize(word.lower(), pos=penn2morphy(tag)) for word, tag in walking_tagged])

['1:14', 'and', 'god', 'say', ',', 'let', 'there', 'be', 'light', 'in', 'the', 'firmament', 'of', 'the', 'heaven', 'to', 'divide', 'the', 'day', 'from', 'the', 'night', ';', 'and', 'let', 'them', 'be', 'for', 'sign', ',', 'and', 'for', 'season', ',', 'and', 'for', 'day', ',', 'and', 'year', ':', '1:15', 'and', 'let', 'them', 'be', 'for', 'light', 'in', 'the', 'firmament', 'of', 'the', 'heaven', 'to', 'give', 'light', 'upon', 'the', 'earth', ':', 'and', 'it', 'be', 'so', '.']


In [163]:
from nltk import pos_tag
from nltk.stem import WordNetLemmatizer

wnl = WordNetLemmatizer()

def penn2morphy(penntag):
    """ Converts Penn Treebank tags to WordNet. """
    morphy_tag = {'NN':'n', 'JJ':'a',
                  'VB':'v', 'RB':'r'}
    try:
        return morphy_tag[penntag[:2]]
    except:
        return 'n' 
    
def lemmatize_sent(text): 
    # Text input is string, returns lowercased strings.
    return [wnl.lemmatize(word.lower(), pos=penn2morphy(tag)) 
            for word, tag in pos_tag(word_tokenize(text))]

print(lemmatize_sent(s8))

['1:14', 'and', 'god', 'say', ',', 'let', 'there', 'be', 'light', 'in', 'the', 'firmament', 'of', 'the', 'heaven', 'to', 'divide', 'the', 'day', 'from', 'the', 'night', ';', 'and', 'let', 'them', 'be', 'for', 'sign', ',', 'and', 'for', 'season', ',', 'and', 'for', 'day', ',', 'and', 'year', ':', '1:15', 'and', 'let', 'them', 'be', 'for', 'light', 'in', 'the', 'firmament', 'of', 'the', 'heaven', 'to', 'give', 'light', 'upon', 'the', 'earth', ':', 'and', 'it', 'be', 'so', '.']


In [164]:
print('Original Single no. 8:')
print(s8, '\n')
print('Lemmatized and removed stopwords:')
print([word for word in lemmatize_sent(s8) 
       if word not in stoplist_combined
       and not word.isdigit() ])

Original Single no. 8:
1:14 And God said, Let there be lights in the firmament of the heaven
to divide the day from the night; and let them be for signs, and for
seasons, and for days, and years: 1:15 And let them be for lights in
the firmament of the heaven to give light upon the earth: and it was
so. 

Lemmatized and removed stopwords:
['1:14', 'god', 'light', 'firmament', 'heaven', 'divide', 'day', 'night', 'sign', 'season', 'day', 'year', '1:15', 'light', 'firmament', 'heaven', 'give', 'light', 'earth']


In [165]:
def preprocess_text(text):
    # Input: str, i.e. document/sentence
    # Output: list(str) , i.e. list of lemmas
    return [word for word in lemmatize_sent(text) 
            if word not in stoplist_combined
            and not word.isdigit()]

In [167]:
from collections import Counter

# Lemmatize and remove stopwords
processed_sent1 = preprocess_text(s8)
processed_sent2 = preprocess_text(single_no9)

In [170]:
print(processed_sent1, '\n', processed_sent2)

['1:14', 'god', 'light', 'firmament', 'heaven', 'divide', 'day', 'night', 'sign', 'season', 'day', 'year', '1:15', 'light', 'firmament', 'heaven', 'give', 'light', 'earth'] 
 ['blonde', 'hair', 'blue', 'eye', 'medium', 'build', 'im', 'early', 'honest', 'care', 'fishing', 'animal', 'golf', 'bike', 'riding', 'tv', 'dvd', 'quiet', 'night', 'home', 'kiss', 'cuddle', 'special', 'person', 'life', 'caring', 'honest', 'lady', 'friendship', 'relationship']


In [171]:
from io import StringIO
from sklearn.feature_extraction.text import CountVectorizer

with StringIO('\n'.join([s8, single_no9])) as fin:
    # Create the vectorizer
    count_vect = CountVectorizer()
    count_vect.fit_transform(fin)

In [ ]:
count_vect.vocabulary_

In [176]:
from io import StringIO
from sklearn.feature_extraction.text import CountVectorizer


with StringIO('\n'.join([s8, single_no9])) as fin:
    # Override the analyzer totally with our preprocess text
    count_vect = CountVectorizer(stop_words=stoplist_combined,
                                 tokenizer=word_tokenize)
    count_vect.fit_transform(fin)
print(count_vect.vocabulary_)

{'1:14': 0, 'god': 21, 'lights': 32, 'firmament': 17, 'heaven': 24, 'divide': 12, 'day': 10, 'night': 35, 'signs': 42, 'seasons': 41, 'days': 11, 'years': 45, '1:15': 1, 'give': 20, 'light': 31, 'earth': 15, 'blonde': 5, 'hair': 23, 'blue': 6, 'eyes': 16, 'medium': 34, 'build': 7, 'im': 27, 'early': 14, '30s': 2, 'honest': 26, 'caring': 8, 'likes': 33, 'fishing': 18, 'animals': 3, 'golf': 22, 'bike': 4, 'riding': 40, 'tv': 44, 'dvds': 13, 'quiet': 38, 'nights': 36, 'home': 25, 'kissing': 28, 'cuddling': 9, 'special': 43, 'person': 37, 'life': 30, 'lady': 29, 'friendship': 19, 'relationship': 39}


/usr/local/lib/python3.7/dist-packages/sklearn/feature_extraction/text.py:401: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ["'d", "'ll", "'m", "'re", "'s", "'ve", '``', 'ai', 'ca', "n't", 'sha', 'wo'] not in stop_words.
  % sorted(inconsistent)


In [177]:
from io import StringIO
from sklearn.feature_extraction.text import CountVectorizer


with StringIO('\n'.join([s8, single_no9])) as fin:
    # Override the analyzer totally with our preprocess text
    count_vect = CountVectorizer(analyzer=preprocess_text)
    count_vect.fit_transform(fin)
print(count_vect.vocabulary_ )

{'1:14': 0, 'god': 20, 'light': 30, 'firmament': 16, 'heaven': 23, 'divide': 11, 'day': 10, 'night': 32, 'sign': 38, 'season': 37, 'year': 41, '1:15': 1, 'give': 19, 'earth': 14, 'blonde': 4, 'hair': 22, 'blue': 5, 'eye': 15, 'medium': 31, 'build': 6, 'im': 26, 'early': 13, 'honest': 25, 'care': 7, 'fishing': 17, 'animal': 2, 'golf': 21, 'bike': 3, 'riding': 36, 'tv': 40, 'dvd': 12, 'quiet': 34, 'home': 24, 'kiss': 27, 'cuddle': 9, 'special': 39, 'person': 33, 'life': 29, 'caring': 8, 'lady': 28, 'friendship': 18, 'relationship': 35}


In [178]:
count_vect.transform([s8, single_no9])

<2x42 sparse matrix of type '<class 'numpy.int64'>'
	with 43 stored elements in Compressed Sparse Row format>

# Walkthrough of Pizza dataset. 
My implementation of IMDB classification is further down, after this pizza classification

In [114]:
!wget https://raw.githubusercontent.com/marycboardman/Random-Acts-of-Pizza/master/train.json
!wget https://raw.githubusercontent.com/marycboardman/Random-Acts-of-Pizza/master/test.json
!wget https://raw.githubusercontent.com/marycboardman/Random-Acts-of-Pizza/master/sampleSubmission.csv

--2022-04-03 01:06:35--  https://raw.githubusercontent.com/marycboardman/Random-Acts-of-Pizza/master/train.json
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.109.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 12569177 (12M) [text/plain]
Saving to: ‘train.json.2’

train.json.2        100%[===================>]  11.99M  --.-KB/s    in 0.09s   

2022-04-03 01:06:35 (138 MB/s) - ‘train.json.2’ saved [12569177/12569177]

--2022-04-03 01:06:35--  https://raw.githubusercontent.com/marycboardman/Random-Acts-of-Pizza/master/test.json
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 301971

In [116]:
!ls

sample_data  test.json	test.json.1  train.json  train.json.1  train.json.2


In [97]:
import json

with open('./train.json') as fin:
    trainjson = json.load(fin)


In [98]:
trainjson[0]

{'giver_username_if_known': 'N/A',
 'number_of_downvotes_of_request_at_retrieval': 0,
 'number_of_upvotes_of_request_at_retrieval': 1,
 'post_was_edited': False,
 'request_id': 't3_l25d7',
 'request_number_of_comments_at_retrieval': 0,
 'request_text': 'Hi I am in need of food for my 4 children we are a military family that has really hit hard times and we have exahusted all means of help just to be able to feed my family and make it through another night is all i ask i know our blessing is coming so whatever u can find in your heart to give is greatly appreciated',
 'request_text_edit_aware': 'Hi I am in need of food for my 4 children we are a military family that has really hit hard times and we have exahusted all means of help just to be able to feed my family and make it through another night is all i ask i know our blessing is coming so whatever u can find in your heart to give is greatly appreciated',
 'request_title': 'Request Colorado Springs Help Us Please',
 'requester_accoun

In [99]:
print('UID:\t', trainjson[0]['request_id'], '\n')
print('Title:\t', trainjson[0]['request_title'], '\n')
print('Text:\t', trainjson[0]['request_text_edit_aware'], '\n')
print('Tag:\t', trainjson[0]['requester_received_pizza'], end='\n')

UID:	 t3_l25d7 

Title:	 Request Colorado Springs Help Us Please 

Text:	 Hi I am in need of food for my 4 children we are a military family that has really hit hard times and we have exahusted all means of help just to be able to feed my family and make it through another night is all i ask i know our blessing is coming so whatever u can find in your heart to give is greatly appreciated 

Tag:	 False


In [100]:
import pandas as pd
df = pd.io.json.json_normalize(trainjson) # Pandas magic... 
df_train = df[['request_id', 'request_title', 
               'request_text_edit_aware', 
               'requester_received_pizza']]
df_train.head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  


request_id                                      request_title  \
0   t3_l25d7            Request Colorado Springs Help Us Please   
1   t3_rcb83  [Request] California, No cash and I could use ...   
2   t3_lpu5j  [Request] Hungry couple in Dundee, Scotland wo...   
3   t3_mxvj3  [Request] In Canada (Ontario), just got home f...   
4  t3_1i6486  [Request] Old friend coming to visit. Would LO...   

                             request_text_edit_aware  requester_received_pizza  
0  Hi I am in need of food for my 4 children we a...                     False  
1  I spent the last money I had on gas today. Im ...                     False  
2  My girlfriend decided it would be a good idea ...                     False  
3  It's cold, I'n hungry, and to be completely ho...                     False  
4  hey guys:\n I love this sub. I think it's grea...                     False

In [102]:

with open('./test.json') as fin:
    testjson = json.load(fin)

In [104]:

print('UID:\t', testjson[0]['request_id'], '\n')
print('Title:\t', testjson[0]['request_title'], '\n')
print('Text:\t', testjson[0]['request_text_edit_aware'], '\n')


UID:	 t3_i8iy4 

Title:	 [request] pregger gf 95 degree house and no food.. promise to pay it forward! Northern Colorado 

Text:	 Hey all! It's about 95 degrees here and our kitchen is pretty much empty save for some bread and cereal.  My girlfriend/fiance is 8 1/2 months pregnant and we could use a good meal.  We promise to pay it forward when we get money! Thanks so much in advance! 



In [105]:
import pandas as pd
df = pd.io.json.json_normalize(testjson)
df_test = df[['request_id', 'request_title', 
               'request_text_edit_aware']]
df_test.head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  


request_id                                      request_title  \
0   t3_i8iy4  [request] pregger gf 95 degree house and no fo...   
1  t3_1mfqi0  [Request] Lost my job day after labour day, st...   
2   t3_lclka                (Request) pizza for my kids please?   
3  t3_1jdgdj  [Request] Just moved to a new state(Waltham MA...   
4   t3_t2qt4  [Request] Two girls in between paychecks, we'v...   

                             request_text_edit_aware  
0  Hey all! It's about 95 degrees here and our ki...  
1  I didn't know a place like this exists! \n\nI ...  
2  Hi Reddit. Im a single dad having a really rou...  
3  Hi I just moved to Waltham MA from my home sta...  
4  We're just sitting here near indianapolis on o...

In [101]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split 

# It doesn't really matter what the function name is called
# but the `train_test_split` is splitting up the data into 
# 2 parts according to the `test_size` argument you've set.

# When we're splitting up the training data, we're spltting up 
# into train, valid split. The function name is just a name =)
train, valid = train_test_split(df_train, test_size=0.2)

In [94]:
# Initialize the vectorizer and 
# override the analyzer totally with the preprocess_text().
# Note: the vectorizer is just an 'empty' object now.
count_vect = CountVectorizer(analyzer=preprocess_text)

# When we use `CounterVectorizer.fit_transform`,
# we essentially create the dictionary and 
# vectorize our input text at the same time.
train_set = count_vect.fit_transform(train['request_text_edit_aware'])
train_tags = train['requester_received_pizza']

# When vectorizing the validation data, we use `CountVectorizer.transform()`.
valid_set = count_vect.transform(valid['request_text_edit_aware'])
valid_tags = valid['requester_received_pizza']

In [106]:
test_set = count_vect.transform(df_test['request_text_edit_aware'])


In [107]:
from sklearn.naive_bayes import MultinomialNB
clf = MultinomialNB() 

# To train the classifier, simple do 
clf.fit(train_set, train_tags) 

MultinomialNB()

In [108]:
from sklearn.metrics import accuracy_score

# To predict our tags (i.e. whether requesters get their pizza), 
# we feed the vectorized `test_set` to .predict()
predictions_valid = clf.predict(valid_set)

print('Pizza reception accuracy = {}'.format(
        accuracy_score(predictions_valid, valid_tags) * 100)
     )

Pizza reception accuracy = 72.52475247524752


In [109]:
count_vect = CountVectorizer(analyzer=preprocess_text)

full_train_set = count_vect.fit_transform(df_train['request_text_edit_aware'])
full_tags = df_train['requester_received_pizza']

# Note: We have to re-vectorize the test set since
#       now our vectorizer is different using the full 
#       training set.
test_set = count_vect.transform(df_test['request_text_edit_aware'])

# To train the classifier
clf = MultinomialNB() 
clf.fit(full_train_set, full_tags) 

MultinomialNB()

In [110]:
# To predict our tags (i.e. whether requesters get their pizza), 
# we feed the vectorized `test_set` to .predict()
predictions = clf.predict(test_set)

In [111]:
success_rate = sum(df_train['requester_received_pizza']) / len(df_train) * 100
print(str('Of {} requests, only {} gets their pizzas,'
          ' {}% success rate...'.format(len(df_train), 
                                        sum(df_train['requester_received_pizza']), 
                                       success_rate)
         )
     )

Of 4040 requests, only 994 gets their pizzas, 24.603960396039604% success rate...


In [112]:
success_rate = sum(predictions) / len(predictions) * 100
print(str('Of {} requests, only {} gets their pizzas,'
          ' {}% success rate...'.format(len(predictions), 
                                        sum(predictions), 
                                       success_rate)
         )
     )

Of 1631 requests, only 51 gets their pizzas, 3.126916002452483% success rate...


# IMDB dataset - Classification

In [181]:
import fastai
from fastai.text import *

In [182]:
#downloading data from fastai
path = untar_data(URLs.IMDB_SAMPLE)

In [183]:
path.ls()

[PosixPath('/root/.fastai/data/imdb_sample/texts.csv')]

In [190]:
#loading the csv data into a pandas dataframe
train_df = pd.read_csv(path/'texts.csv')
train_df = train_df.drop(columns = ['is_valid'])
train_df.head()

label                                               text
0  negative  Un-bleeping-believable! Meg Ryan doesn't even ...
1  positive  This is a extremely well-made film. The acting...
2  negative  Every once in a long while a movie will come a...
3  positive  Name just says it all. I watched this movie wi...
4  negative  This movie succeeds at being one of the most u...

In [199]:
len(train_df)

1000

In [194]:
#splitting train and valid datasets 
train, valid = train_test_split(train_df, test_size=0.2)

In [ ]:
train, valid

In [196]:
# we essentially create the dictionary and 
# vectorize our input text at the same time.
train_set = count_vect.fit_transform(train['text'])
train_tags = train['label']

# When vectorizing the validation data, we use `CountVectorizer.transform()`.
valid_set = count_vect.transform(valid['text'])

valid_tags = valid['label']

In [197]:
#fitting a classifier to the data - Naive Bayes
from sklearn.naive_bayes import MultinomialNB
clf = MultinomialNB() 

# To train the classifier, simple do 
clf.fit(train_set, train_tags) 

MultinomialNB()

In [198]:
# To predict our tags (i.e. whether requesters get their pizza), 
# we feed the vectorized `test_set` to .predict()
predictions_valid = clf.predict(valid_set)

print('Pizza reception accuracy = {}'.format(
        accuracy_score(predictions_valid, valid_tags) * 100)
     )

Pizza reception accuracy = 76.5


As we do not have test set here, I'm considering valid set as final. I got an accuracy score of **76.5%** on the valid set using this Bayes Classification model which is trained on just 800 samples. 